File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:

    void fun(vector<int>& a , int n , int i , vector<vector<int>> &ans , int sum , vector<int> &t , int target){
        if( i == n ){
            return;
        }
        if(sum == target){
            ans.push_back(t);
            return;
        }
        if(sum > target) return;

        t.push_back(a[i]);
        fun(a, n, i, ans, sum + a[i], t, target);
        t.pop_back();
        fun(a, n, i + 1, ans, sum, t, target);
    }
    vector<vector<int>> combinationSum(vector<int>& a, int target) {
        vector<vector<int>> ans;
        int n = a.size();
        vector<int> t ;
        int sum = 0;
        fun(a , n , 0 , ans , sum , t ,  target);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {

    void recu(int[] c, int n, int i, List<List<Integer>> ans, int target, List<Integer> temp) {
        if (target == 0) {
            ans.add(new ArrayList<>(temp));
            return;
        }
        if (i == n || target < 0 ) {
            return;
        }
        // sbse phele add krdo
        temp.add(c[i]);
        // recursion call with same element
        recu(c, n, i, ans, target - c[i], temp);
        // now pop 
        temp.remove(temp.size() - 1);
        // now move forward
        recu(c, n, i + 1, ans, target, temp);

    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        int n = candidates.length;
        recu(candidates, n, 0, ans, target, temp);
        return ans;
    }
}